src/main/java/de/tum/in/www1/artemis/service/WebsocketMessagingService.java

```diff
@@ -109,60 +113,109 @@ public CompletableFuture<Void> sendMessageToUser(String user, String topic, Obje
         }
     }
 
+    /**
+     * Broadcast a new result to the client.
+     * Waits until all notifications are sent and the result properties are restored.
+     * This allows the caller to reuse the passed result object again after calling.
+     *
+     * @param participation used to find the receivers of the notification
+     * @param result        the result object to publish
+     */
+    public void awaitBroadcastNewResult(Participation participation, Result result) {
+        try {
+            // Wait until all notifications got send and the objects were reconnected.
+            broadcastNewResult(participation, result).get();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            log.warn("timed out while sending a result notification", e);
+        }
+    }
+
     /**
      * Broadcast a new result to the client.
      *
      * @param participation the id is used in the destination (so that only clients who have subscribed the specific participation will receive the result)
      * @param result        the new result that should be sent to the client. It typically includes feedback, its participation will be cut off here to reduce the payload size.
      *                          As the participation is already known to the client, we do not need to send it. This also cuts of the exercise (including the potentially huge
      *                          problem statement and the course with all potential attributes
+     * @return a CompletableFuture allowing to wait until all messages got send.
      */
-    public void broadcastNewResult(Participation participation, Result result) {
+    public CompletableFuture<Void> broadcastNewResult(Participation participation, Result result) {
         // remove unnecessary properties to reduce the data sent to the client (we should not send the exercise and its potentially huge problem statement)
         var originalParticipation = result.getParticipation();
         result.setParticipation(originalParticipation.copyParticipationId());
+        List<Result> originalResults = null;
+        if (Hibernate.isInitialized(result.getSubmission())) {
+            var submission = result.getSubmission();
+            submission.setParticipation(null);
+            if (Hibernate.isInitialized(submission.getResults())) {
+                originalResults = submission.getResults();
+                submission.setResults(null);
+            }
+            if (submission instanceof ProgrammingSubmission programmingSubmission && programmingSubmission.isBuildFailed()) {
+                programmingSubmission.setBuildLogEntries(null);
+            }
+        }
```

Comment:
Putting this in a helper method or adding a comment above this block would be helpful. When is the `submission` uninitialized? Why do we need to send it in these cases? (We already talked about this via Slack a bit, but some documentation in the code would be nice)

Reply:
I wanted to put this in a helper method, but we need multiple helper attributes here (originalParticipation, originalResults, originalAssessor, originalFeedback), so I decided to keep it like this.

If the submission is initialized, it also won't be sent to the client.